hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java

```diff
@@ -805,51 +805,73 @@ public boolean isBucketEmpty(String volume, String bucket)
     return deletedKeys;
   }
 
+  /**
+   * @param userName volume owner, null for listing all volumes.
+   * @throws IOException
+   */
   @Override
   public List<OmVolumeArgs> listVolumes(String userName,
       String prefix, String startKey, int maxKeys) throws IOException {
     List<OmVolumeArgs> result = Lists.newArrayList();
-    UserVolumeInfo volumes;
+    List<UserVolumeInfo> volumes = Lists.newArrayList();
+
     if (StringUtil.isBlank(userName)) {
-      throw new OMException("User name is required to list Volumes.",
-          ResultCodes.USER_NOT_FOUND);
+      // null userName represents listing all volumes in cluster.
+      for (String user : getAllUsers()) {
+        volumes.add(getVolumesByUser(user));
+      }
+    } else {
+      volumes.add(getVolumesByUser(userName));
     }
-    volumes = getVolumesByUser(userName);
 
-    if (volumes == null || volumes.getVolumeNamesCount() == 0) {
+    if (volumes.size() == 0) {
       return result;
     }
 
     boolean startKeyFound = Strings.isNullOrEmpty(startKey);
-    for (String volumeName : volumes.getVolumeNamesList()) {
-      if (!Strings.isNullOrEmpty(prefix)) {
-        if (!volumeName.startsWith(prefix)) {
-          continue;
+    for (UserVolumeInfo volInfo : volumes) {
```

Comment:
I think this implementation could be slow if there are many users.  Can the "all volumes" case be handled separately by iterating volumes directly from `getVolumeTable`?  That would avoid: (1) iterating all users, (2) lookup of volumes for each user, (3) lookup of volume by volume name.

Reply:
agreed. To list all Volumes this would be a better way to handle. And leave current code to handle listVolumes for a particular user.

Reply:
Thanks @bharatviswa504 for the review.
I handle the ```listAllVolumes``` separately, and updated.
UT ran well.